From 25f5bd31805bd21d7b7e3583c775252aa8f737e9 Mon Sep 17 00:00:00 2001
From: Dan Prince <dprince@redhat.com>
Date: Wed, 11 Jul 2012 10:19:05 -0400
Subject: [PATCH] Use compute_api.get_all in affinity filters.

Updates the affinity filters so they make a single compute API
call to lookup instance host information rather than single
lookups for each UUID.

This resolves a potential performance issue which can cause a
scheduler to hang while processing requests which contain large numbers
of UUID's in the scheduler_hints.

Fixes LP Bug #1017795.

Change-Id: Ie33378a992e53002c16b2d99135699b576f3d7e3
---
 nova/scheduler/filters/affinity_filter.py |   19 +++++++++++--------
 1 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/nova/scheduler/filters/affinity_filter.py b/nova/scheduler/filters/affinity_filter.py
index e6e7a11..81d9442 100644
--- a/nova/scheduler/filters/affinity_filter.py
+++ b/nova/scheduler/filters/affinity_filter.py
@@ -26,8 +26,11 @@ class AffinityFilter(filters.BaseHostFilter):
     def __init__(self):
         self.compute_api = compute.API()
 
-    def _affinity_host(self, context, instance_id):
-        return self.compute_api.get(context, instance_id)['host']
+    def _all_hosts(self, context):
+        all_hosts = {}
+        for instance in self.compute_api.get_all(context):
+            all_hosts[instance['uuid']] = instance['host']
+        return all_hosts
 
 
 class DifferentHostFilter(AffinityFilter):
@@ -40,9 +43,9 @@ class DifferentHostFilter(AffinityFilter):
 
         affinity_uuids = scheduler_hints.get('different_host', [])
         if affinity_uuids:
-            return not any([i for i
-                              in affinity_uuids
-                              if self._affinity_host(context, i) == me])
+            all_hosts = self._all_hosts(context)
+            return not any([i for i in affinity_uuids
+                              if all_hosts.get(i) == me])
         # With no different_host key
         return True
 
@@ -59,9 +62,9 @@ class SameHostFilter(AffinityFilter):
 
         affinity_uuids = scheduler_hints.get('same_host', [])
         if affinity_uuids:
-            return any([i for i
-                          in affinity_uuids
-                          if self._affinity_host(context, i) == me])
+            all_hosts = self._all_hosts(context)
+            return any([i for i in affinity_uuids
+                          if all_hosts.get(i) == me])
         # With no same_host key
         return True
 
